File: python/users/user_api_solution.py
import os
import boto3
import json

QUEUE_URL = os.getenv("QUEUE_URL", default="test")
sqs = boto3.client('sqs')


# Function to send the message to SQS
def send_message_to_sqs(body):
    print('Sending message to SQS')
    sqs.send_message(
        QueueUrl=QUEUE_URL,
        MessageBody=json.dumps(body)
    )


# Function to validate an email address using a regular expression
def validate_email(email):
    import re
    if re.match(r"[^@]+@[^@]+\.[^@]+", email):
        return True
    return False


# Function to test if a string contains a five digit US zip code
def validate_zip(zip):
    if len(zip) == 5:
        if zip.isdigit():
            return True
    return False


# Function to test if required fields are present in the message body
def validate_required_fields(body):
    if 'username' in body and 'email' in body and 'first_name' in body and 'last_name' in body and 'age' in body and 'city' in body and 'state' in body and 'zip' in body and 'country' in body:
        return True
    return False


# Lambda function to publish user to a queue
def lambda_handler(event, context):
    print(event)
    try:
        body = json.loads(event['body'])
    
        # Validate email
        if not validate_email(body['email']):
            return {'statusCode': 400}
    
        # Validate zip
        if not validate_zip(body['zip']):
            return {'statusCode': 400}
        
        # Validate required fields
        if not validate_required_fields(body):
            return {'statusCode': 400}
    
        send_message_to_sqs(body)
        return {'statusCode': 200}
    except Exception as e:
        print(e)
        return {'statusCode': 500}

File: csharp/src/Users/UserApi_Solution.cs
using System.Net;
using System.Text.RegularExpressions;
using System.ComponentModel.DataAnnotations;

public class UserApi_Solution
{ 
    private readonly IUsersDBContextFactory _usersDBContextFactory;
    
    public UserApi_Solution(IUsersDBContextFactory usersDBContextFactory)
    {
        _usersDBContextFactory = usersDBContextFactory;
    }
    
    //Function to test if Email is not null, not empty and is a valid email
    public bool IsValidEmail(string Email)
    {
        if (Email == null || !new EmailAddressAttribute().IsValid(Email))
        {
            return false;
        }
        return true;
    }

    //Function to test if a string contains a five-digit US zip code
    public bool IsValidZip(string Zip)
    {
        if (Zip == null || !new Regex("^[0-9]{5}$").IsMatch(Zip))
        {
            return false;
        }
        return true;
    }    

    //Function to test if required fields are present in the User object
    public bool IsValidUser(User user)
    {
        if (user.UserName == null || user.Email == null || user.FirstName == null || user.LastName == null || user.Age == 0 || user.City == null || user.State == null || user.Zip == null || user.Country == null)
        {
            return false;
        }
        return true;
    } 

    //Function to validate a new user object for required fields, valid email, valid username, valid zip, save user object and return a RegisterUserResponse object
    public RegisterUserResponse RegisterUser(User user)
    {
        if (!IsValidUser(user))
        {
            return new RegisterUserResponse(HttpStatusCode.BadRequest, "Missing required fields");
        }
        if (!IsValidEmail(user.Email))
        {
            return new RegisterUserResponse(HttpStatusCode.BadRequest, "Invalid email");
        }
        if (!IsValidZip(user.Zip))
        {
            return new RegisterUserResponse(HttpStatusCode.BadRequest, "Invalid zip");
        }
        using (var context = _usersDBContextFactory.CreateDbContext())
        {
            if (context.Users.Any(u => u.UserName == user.UserName))
            {
                return new RegisterUserResponse(HttpStatusCode.BadRequest, "Username already exists");
            }
            context.Users.Add(user);
            context.SaveChanges();
        }
        return new RegisterUserResponse(HttpStatusCode.OK, "User created");
    }

    
}
